### PR TITLE
proto: move extension-specific linking validation into extensions

### PR DESCRIPTION
**Pull request by goaway**
Signed-off-by: Mike Schore <mike.schore@gmail.com>

Description: A couple of the proto descriptors validated during startup are specific to extensions that may be compiled out. This change only validates their presence when the extensions are used. This fixes some spurious validation failures in certain minimal build configurations.
Risk Level: Low
Testing: Ran test suite, ran server.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by goaway** [comment]
I considered moving ratelimit as well, but we didn't actually run into issues with it. (Not sure if it's actually possible to fully compile our ratelimit at this point.) Also an open question whether there are any other cases that should be included. Just wanted to mention in case anyone has insight/opinions.

File: source/extensions/access_loggers/http_grpc/BUILD [hunk 1]
```diff
@@ -31,15 +31,28 @@ envoy_cc_library(
     ],
 )
 
+envoy_cc_library(
+    name = "grpc_access_log_proto_descriptors_lib",
+    srcs = ["grpc_access_log_proto_descriptors.cc"],
+    hdrs = ["grpc_access_log_proto_descriptors.h"],
+    deps = [
+        "//source/common/config:protobuf_link_hacks",
```
Comment (htuch) [inline source/extensions/access_loggers/http_grpc/BUILD]: I don't think the link hacks apply in this situation (see the comment in the respective file).

Comment (goaway) [reply source/extensions/access_loggers/http_grpc/BUILD]: Ah yep, makes sense - this is cruft left over from me copying the dependencies.

**Comment by goaway** [comment]
Not sure what we might be missing - but one thought I floated by Matt in an earlier conversation is that we could also just do dummy linking for everything _instead_ of validation, until whatever needs to be fixed upstream for this all to work is fixed.

**Comment by htuch** [comment]
@goaway I think the challenge is figuring out what "everything" is here. Do you have any thoughts on how that would be computed?

**Comment by goaway** [comment]
Empirically, we encountered just the above two cases when building with only HttpConnMan and the Router as extensions, with a single public entrypoint.

We could possibly induce further failures by building with only a single dummy filter.

However, this really only captures the current state of the world, and further development/refactors could change things. (As an aside, the current validation setup in proto_descriptors.cc will fail fast, but requires a code change to see what actually failed, since it only returns a boolean. We could assert/panic inside that file - though perhaps the boolean return value is better for testing.)

My thought though was merely to replace the validation of each proto currently checked with a dummy ref to ensure it gets compiled in. This won't capture cases we've missed, but it won't reduce our coverage either.

If we really wanted to go all in; we could add a dummy message to any proto file we build in, and explicitly reference those messages whenever the generated header is included.

**Comment by goaway** [comment]
And also, thanks @htuch !

**Comment by htuch** [comment]
@goaway yeah, I guess it would be a bit cleaner, but I'm also wondering if we should just figure out if the suggestion in https://github.com/protocolbuffers/protobuf/issues/4221 makes sense. If we set `always_link`, does it work?

**Comment by goaway** [comment]
@htuch I was under the possible misapprehension that `always_link` didn't work in this case for some reason. If it does, I agree that's cleaner than dummy refs in code.

**Comment by htuch** [comment]
Well, not sure if it does :) The last state in that Protobuf GH issue is that we needed to try it.

**Comment by goaway** [comment]
I don't mind giving it a try. Can at least see if it works for the case we encountered.

**Pull request merged**